### ISSUE TITLE
operator[] with for loop

### ISSUE THREAD

**Issue by user706** [user1095108/ca#1]
Hi,

why do you use a for loop for `operator[]`
https://github.com/user1095108/ca/blob/9622c11f6fc29c1f3a5d8cc6502a128c95592830/array.hpp#L261

Isn't that super-slow for large `i`?

Why not do something like this?
```cpp
  constexpr auto& operator[](size_type i) noexcept
  {
    assert(i < size());
    const auto numTillEnd = N - (f_-a_); return (numTillEnd > i) ? *(f_+i) : *(a_+(i-numTillEnd));
    //const auto numTillEnd = &a_[N-1] - f_; return (numTillEnd >= i) ? *(f_+i) : *(a_+(i-numTillEnd-1));
    //const auto numTillEnd = &a_[N] - f_; return (numTillEnd > i) ? *(f_+i) : *(a_+(i-numTillEnd)); // maybe not safe to take &a_[N] ??
  }
```

Thanks.
(Maybe I'm missing some constexpr magic; but I cannot believe, that constexpr would eliminate a real run-time forloop in this case)

**Comment by user1095108** [comment]
Hey, thanks for the bug! The reason for the for loop is that a circular array, is bidirectional, you usually move one place at a time, so using [] is unusual, but I still wanted to support it, if you wanted to make a quick test, for example. I think your assert is wrong, because you can move endless places in a circular array, but otherwise you are right, operator[] could be made faster, but why write something elaborate for functionality that is unlikely to be used anyway?

**Comment by user706** [comment]
Hi thanks for your comment.

Well I'm actually looking for a ring-buffer with a writer and multiple potential readers, where the readers run delayed; and which has super-fast copying, i.e. not the continuous `CA::S::next` calls when doing `std::copy()`.

You can check some comments about that [here](https://github.com/martinmoene/ring-span-lite/issues/29) in case you're interested, but that code uses completely different iterator, which don't remain fixed to an array position (something I do not like).

I'll be rolling my own ring, but wanted to have a look at some implementations, such as yours, etc.

> I think your assert is wrong, because you can move endless places in a circular array

then you might need an extra (optional) `if`, and `%` operator, to get it right. It is still faster, than potentially thousands of unnecessary loops.

> but why write something elaborate for functionality that is unlikely to be used anyway?

Yes true. But you do not know, if it is unlikely to be used. 
And the for loop in that case is so ridiculous, it just really hurts my eyes (but yes, it's short and neat... but also: super-inefficient)

**Comment by user1095108** [comment]
hmmm, if you're really pedantic, there are issues with pointer arithmetic as well. So, you move size_type places, an unsigned type and if you do pointer arithmetic you get a signed type, it can be + or -, so you lose one bit.

would you like this better?

```
constexpr auto& operator[](size_type i) noexcept
{
  i %= N; auto const n(&a_[N] - f_); return *(i < n ? f_ + i : a_ + (i - n));
}
```

**Comment by user706** [comment]
> ```
> constexpr auto& operator[](size_type i) noexcept
> {
>   i %= N; auto const n(&a_[N] - f_); return *(i < n ? f_ + i : a_ + (i - n));
> }
> ```

It's basically the same as the 2nd commented line from my code [above](https://github.com/user1095108/ca/issues/1#issue-1618698233) at line 6 (except it has the additional `%`). So yes I like it. 
(or hmmm `if (i>=N) i %= N`)

There's a reason I put it in the comment. Is `&a_[N] - f_` safe???
I was kindof looking at this... and wondering... (?)
https://github.com/user1095108/ca/blob/9622c11f6fc29c1f3a5d8cc6502a128c95592830/array.hpp#L20

PS: I think my code above (that is not in comments) is definately safe. right?

**Comment by user1095108** [comment]
You know :), I wrote this a long time ago, so I'm scratching my head as well :) The reason for:

`static_assert(N - 1 <= PTRDIFF_MAX);` 

is that f_ and l_ are always one place apart, if the array is not empty and pointer arithmetic is done on them, and you can take the address of the element one place beyond the end of an array, so I think your code is safe. I'll commit your fix. Otherwise, I'm just using this as a sort of command buffer, and newer commands override the old ones. I tried using std::list and my own list, but a circular array seems faster.

**Comment by user706** [comment]
> I wrote this a long time ago

haha, not so long ago, because it will not run on debian buster, which does not yet have good C++20.

> you can take the address of the element one place beyond the end of an array

Yes, this is I think always true.
So maybe `static_assert(N - 1 <= PTRDIFF_MAX);` is unnecessary, since any array `T a_[N]`, will maybe only accept `N <= PTRDIFF_MAX` anyways.
But I'm just guessing, to be honest...

**Comment by user706** [comment]
Ah:

This...
```cpp
#include <cstdint>

int main()
{
    char aaa[PTRDIFF_MAX-519];
    return 0;
}
```
... compiles.

This
```cpp
#include <cstdint>

int main()
{
    char aaa[PTRDIFF_MAX-518];
    return 0;
}
```
...  fails with
```
3:5: error: total size of local objects 9223372036854775312 exceeds maximum 9223372036854775296
    3 | int main()
      |     ^~~~
```

**Comment by user1095108** [comment]
haha, lol, the reason for the static assert is this:

```
constexpr size_type size() const noexcept
{
  auto const n(l_ - f_); // N - 1 <= PTRDIFF_MAX

  return n < decltype(n){} ? N + n : n;
}
```

so l_ - f_ is at most N - 1 and this needs to be <= PTRDIFF_MAX. As for the debian buster - I always use the unstable packages, so I had c++20 on my raspberry pi box a long time ago, as far as I can remember, really. And I'm using arch on my PC and arch linux always has the latest dev stuff. Anyway, I committed all your stuff now.

**Comment by user706** [comment]
Thanks for your comments!

**Comment by user1095108** [comment]
Don't forget to publish your work. I'll brighten your day with a bug report :)

**Comment by user1095108** [comment]
After thinking about this issue a bit more, I determined that my mental model of a circular array was wrong. A circular array is a random-access container, not a bidirectional-access container. I asked YouChat about this:

> is a circular array a random-access container?
> Yes, a circular array is typically considered to be a random-access container, as it allows for constant-time (O(1)) access to any element of the container using an index. This is because a circular array is essentially a ring-shaped buffer that stores its elements in contiguous memory locations, allowing for efficient indexing into the container. However, because a circular array is a fixed-size data structure that wraps around when it reaches its maximum capacity, it does not support dynamic resizing or insertion/deletion operations in the same way that some other random-access containers (such as a vector) do.

I have made additional code changes because of this, but I'd have to change the iterator as well (make it a random-access iterator).

**Comment by user706** [comment]
Ok great.

I was also missing `operator+=` in the iterator, to be honest...

**Comment by user1095108** [comment]
Ok, now you have everything and more :)

**Comment by user706** [comment]
Nice. I think I'll add a pullrequest for superfast copies, and more sane constructor (`ca::array<int, 4>` for a ring with capacity 3, is maximally confusing (!) and exposes internal design decisions) etc.

By the way: I think the safety here

https://github.com/user1095108/ca/blob/e795289af9b4577e6e8643258eef8d5e843f7fcf/array.hpp#L244-L249

is not sufficient.

You need to add: `static_assert(std::is_same_v<decltype(N + n), decltype(N)>);`

So `l_ - f_` will be at most N-1, since you're design decision is to have the underlying array one larger than the capacity

https://github.com/user1095108/ca/blob/e795289af9b4577e6e8643258eef8d5e843f7fcf/array.hpp#L237

But `N + n`, must automatically convert `decltype(n)` to `decltype(N)` (during the addition) and not the other way around.
Please refer to: https://wandbox.org/permlink/7MPQdBYPa35oIGI7 (or [here](https://wandbox.org/permlink/aOcEZzDqY07FTAiS) for constexpr-version)

Maybe this (`std::is_same_v<decltype(N + n), decltype(N)`) is always the case anyway. But adding 
`static_assert(std::is_same_v<decltype(N + n), decltype(N)>);`
does no harm either and makes it explicit.

**Comment by user1095108** [comment]
circular arrays, circular buffers, ... are really dequeues and I was experimenting with different implementations and I decided to do this one, since it only requires 2 pointers. I only needed a simple container, I don't run any `std::` stuff on my circular array. I'll accept your pull requires and beautify it later, if necessary. As for the static_assert, why bother? There are not that many users of `ca::array` :) If it works, it works.

BTW: This container even runs in dosbox :) I invited you as contributor, but don't be offended, if I beautify your contributions now and then.
